### PR TITLE
[Application] Implement unbonding period

### PR DESCRIPTION
**Pull request by red-0ne**
## Summary

This PR adds an unbonding period when an `Application` submits an unstaking transaction.

* Adds an unbondingHeight to the `Application` proto.
* Resets the unbondingHeight when an `Application` re-stakes during that period.
* Prevents unstaking when an `Application` is already in an unbonding period.
* Adds an EndBlocker routine to unbond `Application`s that reached their unbonding height.
* Deactivate the `Application`, denying it to get/construct `Session`s after the session corresponding to the unstake message ends.

## Issue

`Application`s unstaking needs to be delayed at least until the accounting for the last `Session` is settled.

This is to avoid `Application`s from prematurely unstaking and getting their funds back, before their stake is burned for the work requested.

- #696 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

**Documentation changes** (only if making doc changes)
- [ ] `make docusaurus_start`; only needed if you make doc changes

**Local Testing** (only if making code changes)
- [x] **Unit Tests**: `make go_develop_and_test`
- [x] **LocalNet E2E Tests**: `make test_e2e`
- See [quickstart guide](https://dev.poktroll.com/developer_guide/quickstart) for instructions

**PR Testing** (only if making code changes)
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.
    - **THIS IS VERY EXPENSIVE**, so only do it after all the reviews are complete.
    - Optionally run `make trigger_ci` if you want to re-trigger tests without any code changes
    - If tests fail, try re-running failed tests only using the GitHub UI as shown [here](https://github.com/pokt-network/poktroll/assets/1892194/607984e9-0615-4569-9452-4c730190c1d2)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

## Summary by CodeRabbit

- **New Features**
	- Added the ability to track the end height of unstaking sessions for applications.
	- Introduced a new parameter to configure the number of unbonding sessions applications must wait after unstaking.
  
- **Bug Fixes**
	- Enhanced operational checks in the unstaking process to improve reliability.
  
- **Documentation**
	- Updated test scenarios for clarity regarding application unbonding and staking processes.

- **Tests**
	- Expanded the test suite to cover new unbonding functionality and parameter updates.

### CODE REVIEW

File: e2e/tests/update_params.feature [hunk 1]
```diff
@@ -51,14 +51,15 @@ Feature: Params Namespace
...
-      | supplier_unbonding_period_sessions | 5     | int64 |
+      | name                                  | value | type  |
+      | num_blocks_per_session                | 5     | int64 |
+      | grace_period_end_offset_blocks        | 2     | int64 |
+      | claim_window_open_offset_blocks       | 2     | int64 |
+      | claim_window_close_offset_blocks      | 3     | int64 |
+      | proof_window_open_offset_blocks       | 1     | int64 |
+      | proof_window_close_offset_blocks      | 5     | int64 |
+      | supplier_unbonding_period_sessions    | 5     | int64 |
+      | application_unbonding_period_sessions | 5     | int64 |
```
Comment (Olshansk) [inline e2e/tests/update_params.feature:62]: In the future, let's make governance parameter introductions and business logic changes separate PRs.

File: testutil/keeper/application.go [hunk 2]
```diff
@@ -34,7 +34,9 @@ import (
 // WARNING: Using this map may cause issues if running multiple tests in parallel
 var stakedGatewayMap = make(map[string]struct{})
 
-func ApplicationKeeper(t testing.TB) (keeper.Keeper, context.Context) {
+// ApplicationKeeperWithSharedKeeper creates a new application keeper for testing
+// along with its dependencies returns the ApplicationKeeper, SharedKeeper and context.
+func ApplicationKeeperWithSharedKeeper(t testing.TB) (keeper.Keeper, types.SharedKeeper, context.Context) {
```
Comment (Olshansk) [inline testutil/keeper/application.go:39]: This is not a good pattern and feels easy/lazy.

See the `TokenomicsModuleKeepers` pattern in `testutil/keeper/tokenomics.go` if we really need to embed and reuse different keepers.

File: x/shared/types/params.go [hunk 3]
```diff
@@ -266,6 +299,26 @@ func validateSupplierUnbondingPeriodIsGreaterThanCumulativeProofWindowCloseBlock
 	return nil
 }
 
+// validateApplicationUnbondingPeriodIsGreaterThanCumulativeProofWindowCloseBlocks
+// validates that the ApplicationUnbondingPeriodSession blocks is greater than the cumulative
+// proof window close blocks.
+// It ensures that a supplier cannot unbond before the pending claims are settled.
```
Comment (Olshansk) [inline x/shared/types/params.go:305]:
```suggestion
// ensures that a supplier cannot unbond before the pending claims are settled.
```

File: x/application/keeper/msg_server_unstake_application_test.go [hunk 4]
```diff
@@ -75,6 +152,45 @@ func TestMsgServer_UnstakeApplication_FailIfNotStaked(t *testing.T) {
...
+
+	// Initiate the application unstaking
+	unstakeMsg := &types.MsgUnstakeApplication{Address: appAddr}
+	_, err = srv.UnstakeApplication(ctx, unstakeMsg)
+	require.NoError(t, err)
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	ctx = keepertest.SetBlockHeight(ctx, int64(sdkCtx.BlockHeight()+1))
+
+	_, err = srv.UnstakeApplication(ctx, unstakeMsg)
```
Comment (Olshansk) [inline x/application/keeper/msg_server_unstake_application_test.go:180]:
```suggestion
	// Verify that you cannot unstake if already unstaking.
	_, err = srv.UnstakeApplication(ctx, unstakeMsg)
```

**Review by Olshansk — APPROVED** [review]
Just a couple small nits, but damn, this is great! Clean ✨

File: testutil/keeper/application.go (hunk 2 above)
Comment (red-0ne) [reply testutil/keeper/application.go]: Fixed here dd5e7f19a719be447ca311e2fc209e36d42765e8

**Pull request merged**